dbt_subprojects/dex/macros/models/_project/uniswap_compatible_trades.sql

```diff
@@ -118,4 +118,72 @@ SELECT
     , dexs.evt_index
 FROM
     dexs
+{% endmacro %}
+
+{% macro uniswap_compatible_v4_trades(
+    blockchain = null
+    , project = 'uniswap'
+    , version = '4'
+    , Pair_evt_Swap = source('uniswap_v4_sepolia', 'PoolManager_evt_Swap') 
+    , Factory_evt_PoolCreated = source('uniswap_v4_sepolia', 'PoolManager_evt_Initialize')
+    , taker_column_name = 't.evt_tx_from'
+    , maker_column_name = null
+    , optional_columns = []
+    , pair_column_name = 'f.id'
+    )
+%}
+WITH dexs AS
+(
+    SELECT
+        t.evt_block_number AS block_number
+        , t.evt_block_time AS block_time
+        , {{ taker_column_name }} as taker
+        , {% if maker_column_name -%} {{ maker_column_name }} {% else -%} cast(null as varbinary) {% endif -%} as maker      
+        -- in v4, when amount is negative, then user are selling the token (so things are done from the perspective of the user instead of the pool)
+        , CASE WHEN amount0 < INT256 '0' THEN abs(amount1) ELSE abs(amount0) END AS token_bought_amount_raw 
+        , CASE WHEN amount0 < INT256 '0' THEN abs(amount0) ELSE abs(amount1) END AS token_sold_amount_raw
+        , CASE WHEN amount0 < INT256 '0' THEN f.currency1 ELSE f.currency0 END AS token_bought_address
+        , CASE WHEN amount0 < INT256 '0' THEN f.currency0 ELSE f.currency1 END AS token_sold_address
+        , t.contract_address as project_contract_address
+        {%- if optional_columns %}
+        {%- for optional_column in optional_columns %}
+        , {{ optional_column }}
+        {%- endfor %}
+        {%- endif %}
+        , t.evt_tx_hash AS tx_hash
+        , t.evt_index
+        , row_number() over (partition by t.evt_tx_hash, t.evt_index order by t.evt_block_time desc) as duplicates_rank -- TODO remove after testing
+    FROM
+        {{ Pair_evt_Swap }} t
+    INNER JOIN
+        {{ Factory_evt_PoolCreated }} f
+        ON {{ pair_column_name }} = t.id
+    {%- if is_incremental() %}
+    WHERE
+        {{ incremental_predicate('t.evt_block_time') }}
+    {%- endif %}
+)
+
+SELECT
+    {% if blockchain -%} '{{ blockchain }}' {% else -%} 'Unassigned' {% endif -%} as blockchain
+    , '{{ project }}' AS project
+    , '{{ version }}' AS version
+    , CAST(date_trunc('month', dexs.block_time) AS date) AS block_month
+    , CAST(date_trunc('day', dexs.block_time) AS date) AS block_date
+    , dexs.block_time
+    , dexs.block_number
+    , CAST(dexs.token_bought_amount_raw AS UINT256) AS token_bought_amount_raw
+    , CAST(dexs.token_sold_amount_raw AS UINT256) AS token_sold_amount_raw
+    , dexs.token_bought_address
+    , dexs.token_sold_address
+    , dexs.sender as router
+    , dexs.fee 
+    , dexs.hooks -- if 0x00 null address then no hook, else yes hook
```

Comment:
Key feature for v4, so maybe not in the optional column?
(or optional selection matching `optional_columns` in the param, otherwise null pointer may happen)

Reply:
modified